File: leetcode-04/solutions/python3/0054.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        if not matrix:
            return []

        m = len(matrix)
        n = len(matrix[0])

        ans = [0] * (m * n)
        count = 0
        direction = 0
        up = 0
        down = m - 1
        left = 0
        right = n - 1

        while True:
            if up > down or right < left:
                return ans
            if direction == 0:
                for i in range(left, right + 1):
                    ans[count] = matrix[up][i]
                    count += 1
                up += 1
            elif direction == 1:
                for i in range(up, down + 1):
                    ans[count] = matrix[i][right]
                    count += 1
                right -= 1
            elif direction == 2:
                for i in range(right, left - 1, -1):
                    ans[count] = matrix[down][i]
                    count += 1
                down -= 1
            else:
                for i in range(down, up - 1, -1):
                    ans[count] = matrix[i][left]
                    count += 1
                left += 1
            direction = (direction + 1) % 4

        return ans

File: leetcode-04/solutions/cpp/0054.cpp
class Solution {
 public:
  vector<int> spiralOrder(vector<vector<int>>& matrix) {
    if (matrix.empty()) return {};

    const int m = matrix.size();
    const int n = matrix[0].size();

    vector<int> ans(m * n);
    int count = 0;
    int direction = 0;
    int up = 0;
    int down = m - 1;
    int left = 0;
    int right = n - 1;

    while (true) {
      if (up > down || right < left) return ans;
      if (direction == 0) {
        for (int i = left; i <= right; i++) ans[count++] = matrix[up][i];
        up++;
      } else if (direction == 1) {
        for (int i = up; i <= down; i++) ans[count++] = matrix[i][right];
        right--;
      } else if (direction == 2) {
        for (int i = right; i >= left; i--) ans[count++] = matrix[down][i];
        down--;
      } else {
        for (int i = down; i >= up; i--) ans[count++] = matrix[i][left];
        left++;
      }
      direction = (direction + 1) % 4;
    }

    return ans;
  }
};